File: problems/HackerRank/countingValleys.ts
function countingValleys(steps: number, path: string): number {
  //  8
  // UDDDUDUU
  // /\      _
  //  \    /
  //   \/\/
  // start from 0 and count the number of times coming back 0 (step is Up)

  let stepper = 0;
  let result = 0;

  for (let i = 0; i < steps; i++) {
    const step = path[i];
    step === 'U' ? stepper++ : stepper--;

    // go back to sea level
    if (stepper === 0 && i !== 0) {
      step === 'U' && result++;
    }
  }
  return result;
}

console.log(countingValleys(10, 'UDDDUDUUDU'));

File: problems/HackerRank/week2/countingValleys.py
def countingValleys(steps, path):
    
    ans = 0
    level = 0
    for ele in path:
        
        if ele == "U":
            level += 1
        elif ele == "D":
            level -= 1
    
        if level == 0 and ele == "U":
            ans += 1
    
    return ans